### PR TITLE
bpo-42725: Render annotations effectless on symbol table with PEP 563

### PR DESCRIPTION
**Pull request by isidentical**
https://bugs.python.org/issue42725

### CODE REVIEW

File: Lib/test/test_future.py [hunk 1]
```diff
@@ -310,19 +318,13 @@ def test_annotations(self):
         eq("f'{x}'")
         eq("f'{x!r}'")
         eq("f'{x!a}'")
-        eq('(yield from outside_of_generator)')
```
Comment (pablogsal) [inline Lib/test/test_future.py:313]: Can we move all these expressions to `test_annotations_forbidden` so is clear that we are not missing any case?

File: Misc/NEWS.d/next/Core and Builtins/2021-04-25-05-40-51.bpo-42725.WGloYm.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Usage of ``await``/``yield``/``yield from`` and named expressions within an
+annotation is now forbidden when PEP 563 is activated. Also the annotations
+won't interact with symbol table at all, and will be considered no different
+than strings.
```
Comment (pablogsal) [inline Misc/NEWS.d/next/Core and Builtins/2021-04-25-05-40-51.bpo-42725.WGloYm.rst]:
```suggestion
Usage of ``await``/``yield``/``yield from`` and named expressions within an
annotation is now forbidden when PEP 563 is activated.
```

The users don't need to know what is a symbol table :)

File: Python/symtable.c [hunk 3]
```diff
@@ -45,6 +45,14 @@
 #define NAMED_EXPR_COMP_ITER_EXPR \
 "assignment expression cannot be used in a comprehension iterable expression"
 
+#define ANNOTATION_NOT_ALLOWED \
+"'%s' can not be used within an annotation"
+
+// Annotation blocks shouldn't have any affect on the symbol table
+// since in the compilation stage, they will all be transformed to
+// strings.
+#define IS_PSUEDO_BLOCK(BLOCK) (BLOCK == AnnotationBlock)
```
Comment (pablogsal) [inline Python/symtable.c]: `IS_PSUEDO_BLOCK` is quite confusing. Let's name this IS_ANNOTATION_BLOCK or something along those lines

File: Python/symtable.c [hunk 4]
```diff
@@ -2001,6 +2055,20 @@ symtable_visit_dictcomp(struct symtable *st, expr_ty e)
                                          e->v.DictComp.value);
 }
 
+static int
+symtable_check_annotation(struct symtable *st, const char *name, expr_ty e)
```
Comment (pablogsal) [inline Python/symtable.c]: This name is quite non descriptive, let's call this something like `symtable_raise_if_annotation_block` or something like that.

File: Python/symtable.c [hunk 5]
```diff
@@ -2001,6 +2055,20 @@ symtable_visit_dictcomp(struct symtable *st, expr_ty e)
...
+    if (st->st_cur->ste_type == AnnotationBlock) {
+        PyErr_Format(PyExc_SyntaxError, ANNOTATION_NOT_ALLOWED, name);
+        PyErr_RangedSyntaxLocationObject(st->st_filename,
+                                         e->lineno,
+                                         e->col_offset + 1,
+                                         e->end_lineno,
+                                         e->end_col_offset + 1);
+        return 0;
+    }
+    return 1;
```
Comment (pablogsal) [inline Python/symtable.c]:
```suggestion
    if (st->st_cur->ste_type != AnnotationBlock) {
        return 1;
    }
    PyErr_Format(PyExc_SyntaxError, ANNOTATION_NOT_ALLOWED, name);
    PyErr_RangedSyntaxLocationObject(st->st_filename,
                                     e->lineno,
                                     e->col_offset + 1,
                                     e->end_lineno,
                                     e->end_col_offset + 1);
    return 0;
```

File: Modules/symtablemodule.c [hunk 6]
```diff
@@ -58,7 +58,6 @@ _symtable_symtable_impl(PyObject *module, PyObject *source,
     }
     t = (PyObject *)st->st_top;
     Py_INCREF(t);
-    PyMem_Free((void *)st->st_future);
```
Comment (pablogsal) [inline Modules/symtablemodule.c:61]: Aren't you leaking `st_future`? Who is freeing this memory now?

**Review by NAPADAI — APPROVED** [review]

File: Modules/symtablemodule.c (hunk 6 above)
Comment (isidentical) [reply Modules/symtablemodule.c:61]: This seems like something we caught while testing, though when you create a symbol table from a string (for python interface), then it doesn't actually set the futures. And while checking that, I noticed that this free was unnecessary and previously had no effect. Now it is trying to deallocate something that we didn't allocate in this function in the first place.  Moving the `PyMem_Free` call to the host side (`_PySymtableStringObjectFlags`).

**Comment by isidentical** [comment]
```
 $ ./python -m test -R3:3 test_symtable test_future
0:00:00 load avg: 2.61 Run tests sequentially
0:00:00 load avg: 2.61 [1/2] test_symtable
beginning 6 repetitions
123456
......
0:00:00 load avg: 2.61 [2/2] test_future
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

All 2 tests OK.

Total duration: 565 ms
Tests result: SUCCESS
```

**Pull request merged**